File: app/src/main/java/com/bongorian/signa1/FaultConfig.java
package com.bongorian.signa1;

/** Immutable LIVE direction and optional measured-input coupling. */
final class FaultConfig {
    final boolean enabled,motion,audio,timing,thermal,cpu;final float sensitivity;final int mains;final LivePerformance performance;
    FaultConfig(boolean enabled,boolean motion,boolean audio,boolean timing,boolean thermal,boolean cpu,float sensitivity,int mains){this(enabled,motion,audio,timing,thermal,cpu,sensitivity,mains,LivePerformance.defaults());}
    FaultConfig(boolean enabled,boolean motion,boolean audio,boolean timing,boolean thermal,boolean cpu,float sensitivity,int mains,LivePerformance performance){this.performance=performance;this.enabled=enabled;this.motion=motion;this.audio=audio;this.timing=timing;this.thermal=thermal;this.cpu=cpu;this.sensitivity=EffectParameters.unit(sensitivity);this.mains=mains==60?60:50;}
    static FaultConfig defaults(){return new FaultConfig(false,true,false,true,true,true,.5f,50);}
    FaultConfig enabled(boolean value){return new FaultConfig(value,motion,audio,timing,thermal,cpu,sensitivity,mains,performance);}
    FaultConfig audio(boolean value){return new FaultConfig(enabled,motion,value,timing,thermal,cpu,sensitivity,mains,performance);}
    FaultConfig performance(LivePerformance value){return new FaultConfig(enabled,motion,audio,timing,thermal,cpu,sensitivity,mains,value);}
}

File: app/src/main/kotlin/com/bongorian/signa1/FaultConfig.kt
package com.bongorian.signa1

/** Immutable LIVE direction and optional measured-input coupling. */
internal class FaultConfig
constructor(
    val enabled: Boolean,
    val motion: Boolean,
    val audio: Boolean,
    val timing: Boolean,
    val thermal: Boolean,
    val cpu: Boolean,
    sensitivity: Float,
    mains: Int,
    val performance: LivePerformance = LivePerformance.defaults(),
    val experimental: Boolean = false,
) {
    val sensitivity: Float
    val mains: Int

    init {
        this.sensitivity = EffectParameters.unit(sensitivity)
        this.mains = if (mains == 60) 60 else 50
    }

    fun enabled(value: Boolean): FaultConfig {
        return FaultConfig(
            value,
            motion,
            audio,
            timing,
            thermal,
            cpu,
            sensitivity,
            mains,
            performance,
            experimental,
        )
    }

    fun audio(value: Boolean): FaultConfig {
        return FaultConfig(
            enabled,
            motion,
            value,
            timing,
            thermal,
            cpu,
            sensitivity,
            mains,
            performance,
            experimental,
        )
    }

    fun performance(value: LivePerformance): FaultConfig {
        return FaultConfig(enabled, motion, audio, timing, thermal, cpu, sensitivity, mains, value, experimental)
    }

    fun experimental(value: Boolean): FaultConfig =
        if (value == experimental) this else FaultConfig(enabled, motion, audio, timing, thermal, cpu, sensitivity, mains, performance, value)

    companion object {
        fun defaults(): FaultConfig {
            return FaultConfig(false, true, false, true, true, true, .5f, 50)
        }
    }
}
